Add support for managed task inputs

eatkins opened the pull request:
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) guidelines

This PR adds support for `make`-like file dependencies. sbt will now track the output files of tasks that return `Seq[File]`, `File`, `Seq[Path]`, `Path`. Other tasks can then depend on these tasks and avoid re-running the dependent task if the output paths of the parent tasks have not changed. Moreover, the input files that are specified using the fileInputs key are automatically tracked in a continuous build.

To illustrate the power of this new functionality, I added a scripted test nio/make-clone that demonstrates the make like dependency management.

I copied the scripted test to a different console and ran a local version of sbt published off this build. I then ran the following commands:

```
sbt:make-clone> ~executeMain 1 2 3
[info] Not compiling libfoo: no inputs have changed.
[info] Not running linker: no outputs have changed.
[info] Not building /Users/ethanatkins/work/scratch/make-clone/out/main/main.out: no dependencies have changed
[info] Running /Users/ethanatkins/work/scratch/make-clone/out/main/main.out with arguments: '1', '2', '3'
[info] f := x * x * x
[info] f(1) = 1
[info] f(2) = 8
[info] f(3) = 27
Total time: 21 ms
  make-clone / executeMain                    : 9 ms
  make-clone / compileLib                     : 2 ms
  make-clone / linkLib                        : 1 ms
  make-clone / compileMain                    : 1 ms
  make-clone / compileLib / changedOutputPaths: 1 ms
[success] Total time: 0 s, completed Apr 28, 2019 10:40:14 PM
[info] 1. Monitoring source files for updates...
[info] Project: make-clone
[info] Command: executeMain 1 2 3
[info] Options:
[info]   <enter>: return to the shell
[info]   'r': repeat the current command
[info]   'x': exit sbt

// I rewrote the file lib.c to change the function definition from `x * x * x` to `x * x` to trigger a build

[info] Build triggered by /Users/ethanatkins/work/scratch/make-clone/src/lib/lib.c. Running 'executeMain 1 2 3'.
[info] Compiling /Users/ethanatkins/work/scratch/make-clone/src/lib/lib.c to /Users/ethanatkins/work/scratch/make-clone/out/lib/objects/lib.o
[info] Linking /Users/ethanatkins/work/scratch/make-clone/out/lib/libfoo.dylib
[info] Building executable /Users/ethanatkins/work/scratch/make-clone/out/main/main.out
[info] Running /Users/ethanatkins/work/scratch/make-clone/out/main/main.out with arguments: '1', '2', '3'
[info] f := x * x
[info] f(1) = 1
[info] f(2) = 4
[info] f(3) = 9
Total time: 250 ms
  make-clone / compileMain                 : 101 ms
  make-clone / compileLib                  :  73 ms
  make-clone / linkLib                     :  54 ms
  make-clone / executeMain                 :   8 ms
  make-clone / linkLib / changedOutputPaths:   1 ms
[success] Total time: 0 s, completed Apr 28, 2019 10:40:25 PM
[info] 2. Monitoring source files for updates...
[info] Project: make-clone
[info] Command: executeMain 1 2 3
[info] Options:
[info]   <enter>: return to the shell
[info]   'r': repeat the current command
[info]   'x': exit sbt

// Here I just touched the lib.c file

[info] Build triggered by /Users/ethanatkins/work/scratch/make-clone/src/lib/lib.c. Running 'executeMain 1 2 3'.
[info] Not compiling libfoo: no inputs have changed.
[info] Not running linker: no outputs have changed.
[info] Not building /Users/ethanatkins/work/scratch/make-clone/out/main/main.out: no dependencies have changed
[info] Running /Users/ethanatkins/work/scratch/make-clone/out/main/main.out with arguments: '1', '2', '3'
[info] f := x * x
[info] f(1) = 1
[info] f(2) = 4
[info] f(3) = 9
Total time: 25 ms
  make-clone / executeMain: 9 ms
  make-clone / compileMain: 1 ms
  make-clone / compileLib : 1 ms
  make-clone / linkLib    : 1 ms
[success] Total time: 0 s, completed Apr 28, 2019 10:40:26 PM
[info] 3. Monitoring source files for updates...
[info] Project: make-clone
[info] Command: executeMain 1 2 3
[info] Options:
[info]   <enter>: return to the shell
[info]   'r': repeat the current command
[info]   'x': exit sbt
sbt:make-clone> compileLib/clean
Total time: 30 ms
  make-clone / compileLib / clean: 8 ms
[success] Total time: 0 s, completed Apr 28, 2019 10:40:41 PM
sbt:make-clone> executeMain 4 5 6
[info] Compiling /Users/ethanatkins/work/scratch/make-clone/src/lib/lib.c to /Users/ethanatkins/work/scratch/make-clone/out/lib/objects/lib.o

// notice that compileLib / clean did force a rebuild of the lib.c file

[info] Linking /Users/ethanatkins/work/scratch/make-clone/out/lib/libfoo.dylib
[info] Building executable /Users/ethanatkins/work/scratch/make-clone/out/main/main.out
[info] Running /Users/ethanatkins/work/scratch/make-clone/out/main/main.out with arguments: '4', '5', '6'
[info] f := x * x
[info] f(4) = 16
[info] f(5) = 25
[info] f(6) = 36
Total time: 239 ms
  make-clone / compileMain: 100 ms
  make-clone / compileLib :  75 ms
  make-clone / linkLib    :  50 ms
  make-clone / executeMain:   8 ms
[success] Total time: 0 s, completed Apr 28, 2019 10:40:46 PM
sbt:make-clone> linkLib / clean
Total time: 34 ms
  make-clone / linkLib / clean: 6 ms
[success] Total time: 0 s, completed Apr 28, 2019 10:48:53 PM
sbt:make-clone> executeMain 1 2 3
[info] Not compiling libfoo: no inputs have changed.

// Notice that running linkLib / clean did not invalidate lib.o but did invalidate everything else in the chain

[info] Linking /Users/ethanatkins/work/scratch/make-clone/out/lib/libfoo.dylib
[info] Building executable /Users/ethanatkins/work/scratch/make-clone/out/main/main.out
[info] Running /Users/ethanatkins/work/scratch/make-clone/out/main/main.out with arguments: '1', '2', '3'
[info] f := x * x
[info] f(1) = 1
[info] f(2) = 4
[info] f(3) = 9
Total time: 235 ms
  make-clone / compileMain                            : 108 ms
  make-clone / linkLib                                :  92 ms
  make-clone / executeMain                            :   8 ms
  make-clone / compileLib                             :   2 ms
  make-clone / linkLib / changedOutputPaths           :   1 ms
  make-clone / compileMain / inputFileStamps          :   1 ms
  make-clone / compileLib / changedOutputPaths        :   1 ms
  make-clone / compileLib / allInputPathsAndAttributes:   1 ms
  make-clone / compileLib / changedInputFiles         :   1 ms
[success] Total time: 0 s, completed Apr 28, 2019 10:48:58 PM
```

eatkins committed:
Bump io

The newest version of io repackages a number of classes into the
sbt.nio.* packages. It also changes some of the semantics of glob
related apis. This commit updates all of the usages of the updated apis
within sbt but should have no functional difference.

eatkins committed:
Undeprecate WatchSource

Since the new watch implementation has yet to be widely deployed, we
should hold off on deprecating the old keys. They could still be
deprecated in a patch release or in 1.4.0.

eatkins committed:
Remove unneeded filters

eatkins committed:
Make Repository private[sbt]

This trait may not even survive until 1.4.0. It should definitely not be
public. I got a little overexcited about programming with higher kinded
types when I added it.

eatkins committed:
Remove duped system in

I had some ideas for allowing the user to get a copy of system in during
a continuous build but I can't really see a good use case now so I'm
going to remove it before 1.3.0.

eatkins committed:
Clean up nio apis

This commit refactors things so that the nio apis are located primarily
in the nio package. Because the nio keys are a first class sbt feature,
I had to add import sbt.nio._ and sbt.nio.Keys._ to the autoimports in
BuildUtil.scala

eatkins committed:
Remove unneeded test

This isn't testing anything useful after the refactoring that removed
the implicit FileTree.Repository.

eatkins committed:
Remove pointless test

I never actually implemented this test so it wasn't really testing
anything.

eatkins committed:
Add support for managed task inputs

In my recent changes to watch, I have been moving towards a world in
which sbt manages the file inputs and outputs at the task level. The
main idea is that we want to enable a user to specify the inputs and
outputs of a task and have sbt able to track those inputs across
multiple task evaluations. Sbt should be able to automatically trigger a
build when the inputs change and it also should be able to avoid task
evaluation if non of the inputs have changed.

The former case of having sbt automatically watch the file inputs of a
task has been present since watch was refactored. In this commit, I
make it possible for the user to retrieve the lists of new, modified and
deleted files. The user can then avoid task evaluation if none of the
inputs have changed.

To implement this, I inject a number of new settings during project
load if the fileInputs setting is defined for a task. The injected
settings are:

allPathsAndAttributes -- this retrieves all of the paths described by
  the fileInputs for the task along with their attributes
fileStamps -- this retrieves all of the file stamps for the files
  returned by allPathsAndAttributes

Using these two injected tasks, I also inject a number of derived tasks,
such as allFiles, which returns all of the regular files returned by
allPathsAndAttributes and changedFiles, which returns all of the regular
files that have been modified since the last run.

Using these injected settings, the user is able to write tasks that
avoid evaluation if the inputs haven't changed.

foo / fileInputs += baseDirectory.value.toGlob / ** / "*.scala"
foo := {
  foo.previous match {
    case Some(p) if (foo / changedFiles).value.isEmpty => p
    case _ => fooImpl((foo / allFiles).value
  }
}

To make this whole mechanism work, I add a private task key:
val fileAttributeMap = taskKey[java.util.HashMap[Path, Stamp]]("...")
This keeps track of the stamps for all of the files that are managed by
sbt. The fileStamps task will first look for the stamp in the attribute
map and, only if it is not present, it will update the cache. This
allows us to ensure that a given file will only be stamped once per task
evaluation run no matter how the file inputs are specified. Moreover, in
a continuous build, I'm able to reuse the attribute map which can
significantly reduce latency because the default file stamping
implementation used by zinc is fairly expensive (it can take anywhere
between 300-1500ms to stamp 5000 8kb source files on my mac).

I also renamed some of the watch related keys to be a bit more clear.

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1553/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1553/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@a5ab32c775af39cd44a8b97fee2c9f7b70588d30
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@dc94df28346eae2a6d1486ba59be6aeb96d9c3ce
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@eae491f96e51670dfb8b2b3fd516bdb43d55692d
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

eatkins committed:
Add full support for managing file task io

This commit unifies my previous work for automatically watching the
input files for a task with support for automatically tracking and
cleaning up the output files of a task. The big idea is that users may
want to define tasks that depend on the file outputs of other tasks and
we may not want to run the dependent tasks if the output files of the
parent tasks are unmodified.

For example, suppose we wanted to make a plugin for managing typescript
files. There may be, say, two tasks with the following inputs and
outputs:

compileTypescript = taskKey[Unit]("shells out to compile typescript files")
  fileInputs -- sourceDirectory / ** / "*.ts"
  fileOutputs -- target / "generated-js" / ** / "*.js"

minifyGeneratedJS = taskKey[Path]("minifies the js files generated by compileTypescript to a single combined js file.")
  dependsOn: compileTypeScript / fileOutputs

Given a clean build, the following should happen
> minifyGeneratedJS
// compileTypescript is run
// minifyGeneratedJS is run

> minifyGeneratedJS
// no op because nothing changed

> minifyGeneratedJS / clean
// removes the file returned by minifyGeneratedJS.previous

> minifyGeneratedJS
// re-runs minifyGeneratedJS with the previously compiled js artifacts

> compileTypescript / clean
// removes the generated js files

> minifyGeneratedJS
// compileTypescript is run because the previous clean removed the generated js files
// minifyGeneratedJS runs because the artifacts have changed

> clean
// removes the generated js files and the minified js file

> minifyGeneratedJS
// compileTypescript is run because the generated js files were
// minifyGeneratedJS is run both because it was removed and

Moreover, if compileTypescript fails, we want minifyGeneratedJS to fail
as well.

This commit makes this all possible. It adds a number of tasks to
sbt.nio.Keys that deal with the output files. When injecting settings, I
now identify all tasks that return Seq[File], File, Seq[Path] and Path
and create a hidden special task: dynamicFileOutputs: TaskKey[Seq[Path]]
This special task runs the underlying task and converts the result to
Seq[Path]. From there, we can have the tasks like changedOutputPaths
delegate to dynamicFileOutputs which, by proxy, runs the underlying
task. If any task in the input / output chain fails, the entire sequence
fails.

Unlike the fileInputs, we do not register the dynamicFileOutputs or
fileOutputs with continuous watch service so these paths will not
trigger a continuous build if they are modified. Only explicit unmanaged
input sources should should do that.

As part of this, I also added automatic generation of a custom clean task for
any task that returns Seq[File], File, Seq[Path] or Path. I also added
aggregation so that clean can be defined in a configuration or project
and it will automatically run clean for all of the tasks that have a
custom clean implementation in that task or project. The automatic clean
task will only delete files that are in the task target directory to
avoid accidentally deleting unmanaged files.

eatkins committed:
Rename InputGraph to SettingsGraph

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1567/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1567/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@a5ab32c775af39cd44a8b97fee2c9f7b70588d30
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@eae491f96e51670dfb8b2b3fd516bdb43d55692d
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1568/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1568/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@a5ab32c775af39cd44a8b97fee2c9f7b70588d30
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@eae491f96e51670dfb8b2b3fd516bdb43d55692d
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@f073dde812848ef443619578f55e290f3863bcaa
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1569/console">is in progress...</a>

eatkins commented:
I need to fix a few scripted tests tomorrow and add a few commit messages, but I'm otherwise very happy with this.

eatkins commented on sbt/src/sbt-test/nio/make-clone/project/build.properties:
```diff
@@ -0,0 +1 @@
+sbt.version=1.3.0-pre-rc-6
```
Whoops. This file will be removed.

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1569/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@a5ab32c775af39cd44a8b97fee2c9f7b70588d30
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@eae491f96e51670dfb8b2b3fd516bdb43d55692d
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@18f4fc75e5b307bb8afef49865be23cff57c9111
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

eed3si9n commented on build.sbt:
```diff
@@ -11,6 +11,7 @@ import scala.xml.transform.{RewriteRule, RuleTransformer}
 def buildLevelSettings: Seq[Setting[_]] =
   inThisBuild(
     Seq(
+      doc in Compile := file("/dev/null"),
```
This should probably go.

eatkins replied:
Yes. I keep accidentally pushing the commit that adds that.

eed3si9n commented on sbt/src/sbt-test/nio/clean/build.sbt:
```diff
@@ -0,0 +1,37 @@
+
+val copyFile = taskKey[Int]("dummy task")
+copyFile / fileInputs += baseDirectory.value.toGlob / "base" / "*.txt"
+copyFile / fileOutputs += baseDirectory.value.toGlob / "out" / "*.txt"
```
Does registering `fileOutputs` automatically invalidate the cache?

eatkins replied:
I'm not sure what you mean by the cache. The fileOutputs setting is used in two tasks: the custom clean implementation and the allOutputPaths task defined in sbt.nio.Keys. The allOutputPaths task writes to the previous cache so that we can implement the modifiedOutputPaths, changedOutputPaths, removedOutputPaths and outputFileStamps tasks.

Also both fileInputs and fileOutputs are settings so there shouldn't be side effects in running them.

eed3si9n replied:
I was thinking of a situation where you have task `foo` whose `foo / fileInputs` point to `baseDirectory.value.toGlob / "out" / "*.txt"`. When `copyFile` runs, would that automatically invalidate the internal stamps, or would `foo` need to poll, or explicitly depend on `copyFile`?

eatkins replied:
This test should be rewritten so that copyFile returns Seq[Path]. Presently, a dependent task would implicitly poll `baseDirectory.value.toGlob / "out" / "*.txt"` if it called `(copyFile / allOutputPaths).value`. If instead, copyFile returned `Seq[Path]` then `fileOutputs` would not need to be specified at all and `(copyFile / allOutputPaths).value` would return the paths from the previous cache for `copyFile.value`.

eatkins replied:
I hadn't yet thought of automatically caching the values of a task returning a collection of files or paths when I wrote this test.

The situation where fileOutputs is useful is when your task delegates to an external task that has the side effect of writing files into an output path but doesn't return the output files. For example if you shell out to invoke the typescript compiler on a set of files, you don't necessarily know what all the output files are but you may know what directory they will be in and what extension they have. So you might do
```
typeScript / fileOutputs += target.value.toGlob / "js" / ** / "*.js"
```
and then you might have another task the minifies that looks like and delegates to some function `def minify(minifiedJS: Path, inputs: Seq[Path]): Path` where the return value is the generated javascript file, we could write
```scala
minifyJS := {
  val allPaths = (typeScript / allOutputPaths.value)
  val outputFile = target.value.toPath
  minifyJS.previous match {
    case Some(o: Path) if (typeScript / changedOutputPaths).value.isEmpty => o
    case _ => minify(outputFile, allPaths)
  }
 }
```
The `(typeScript / allOutputPaths).value` will proxy through (typeScript / fileOutputs). This will poll the file system. The `(typeScript / changedOutputPaths).value` will delegate to `(typeScript / outputFileStamps).value and `(typeScript / outputFileStamps).previous`. The implementation of outputFileStamps delegates to allOutputPaths, so we only ever poll `baseDirectory.value.toGlob / "out" / "*.txt"` once.

eatkins replied:
If `typeScript := taskKey[Seq[Path]]("Compiles input typescript files and returns the generated js files")`, then there would be no reason to define `typeScript / fileOutputs`.

eatkins replied:
Upon reflection I realized there was an issue with the implementation. I updated the PR so that if you call (copyFile / allOutputPaths).value, it will invoke copyFile. Previously this was only the case if copyFile returned a File, Path or Seq of one or the other. You can see this by inspecting the graph:

```
sbt:nio-clean> inspect copyFile / allOutputPaths
[info] Task: scala.collection.Seq[java.nio.file.Path]
[info] Description:
[info]  All of the file output for a task with no filters applied.
[info] Provided by:
[info]  ProjectRef(uri("file:/Users/ethanatkins/work/scratch/nio-clean/"), "nio-clean") / copyFile / allOutputPaths
[info] Defined at:
[info]  (sbt.nio.Settings.allOutputPathsImpl) Settings.scala:286
[info] Dependencies:
[info]  copyFile / dynamicFileOutputs
[info]  ./ ./ ./ fileTreeView
[info]  copyFile / fileOutputs
[info] Reverse dependencies:
[info]  copyFile / outputFileStamps
[info] Delegates:
[info]  copyFile / allOutputPaths
[info]  allOutputPaths
[info]  ThisBuild / copyFile / allOutputPaths
[info]  ThisBuild / allOutputPaths
[info]  Zero / copyFile / allOutputPaths
[info]  Global / allOutputPaths
[info] Related:
[info]  Test / package / allOutputPaths
[info]  Test / packageSrc / allOutputPaths
[info]  Compile / packageSrc / allOutputPaths
[info]  Runtime / dependencyDot / allOutputPaths
[info]  Test / products / allOutputPaths
[info]  Test / productDirectories / allOutputPaths
[info]  Runtime / ivyReport / allOutputPaths
[info]  Provided / dependencyGraphMl / allOutputPaths
[info]  Test / ivyReport / allOutputPaths
[info]  Optional / ivyReport / allOutputPaths
[info] ...
sbt:nio-clean> inspect copyFile / dynamicFileOutputs
[info] Task: scala.collection.Seq[java.nio.file.Path]
[info] Description:
[info]  The outputs of a task
[info] Provided by:
[info]  ProjectRef(uri("file:/Users/ethanatkins/work/scratch/nio-clean/"), "nio-clean") / copyFile / dynamicFileOutputs
[info] Defined at:
[info]  /Users/ethanatkins/work/scratch/nio-clean/build.sbt:7
[info] Dependencies:
[info]  copyFile
[info] Reverse dependencies:
[info]  copyFile / allOutputPaths
[info] Delegates:
[info]  copyFile / dynamicFileOutputs
[info]  dynamicFileOutputs
[info]  ThisBuild / copyFile / dynamicFileOutputs
[info]  ThisBuild / dynamicFileOutputs
[info]  Zero / copyFile / dynamicFileOutputs
[info]  Global / dynamicFileOutputs
[info] Related:
[info]  Provided / ivyReport / dynamicFileOutputs
[info]  Compile / productDirectories / dynamicFileOutputs
[info]  Test / managedResources / dynamicFileOutputs
[info]  Test / products / dynamicFileOutputs
[info]  Test / doc / dynamicFileOutputs
[info]  Compile / package / dynamicFileOutputs
[info]  Compile / doc / dynamicFileOutputs
[info]  Compile / packageBin / dynamicFileOutputs
[info]  Compile / products / dynamicFileOutputs
[info]  Test / resources / dynamicFileOutputs
[info] ...
sbt:nio-clean> 
```
In particular note that copyFile / dynamicFileOutputs has a dependency on copyFile.

eatkins committed:
Add disabled to list of valid scripted test names

I noticed that when I ran scripted locally (with no arguments) it was
hard to tell if they succeeded or not because the last test it ran was
metadata-only-resolver which had no valid test files and dumped a bunch
of lines to stderr. I noticed there were a number of tests that had
files named disabled but no other tests with a test named disable so I
renamed metadata-only-resolver to disabled..

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1572/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1572/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@7e80be7a020320252085695f39f93e9bcf0fdb42
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@a365362e35d56e0b578b2f6b8455f068169a8a57
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@18f4fc75e5b307bb8afef49865be23cff57c9111
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1573/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1573/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@7e80be7a020320252085695f39f93e9bcf0fdb42
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@a365362e35d56e0b578b2f6b8455f068169a8a57
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@18f4fc75e5b307bb8afef49865be23cff57c9111
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

eatkins commented:
~~I ran the scripted tests locally on my machine and there are 4 failing tests as of the current revision. I should be able to fix them without too much trouble.~~

Nevermind. Fixed them.

eatkins committed:
Add package/mappings to launcher base scripted tests

I don't understand why this wasn't failing before.

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1574/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1574/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@9bf8076e22f30589942d0ccb1baf9d0abd7003b0
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@d1c6f0b1487f90f107e9fd954f52daa15b6c373a
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@18f4fc75e5b307bb8afef49865be23cff57c9111
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1575/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1575/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@58c377e943040d94f8b831c85ca1e91d4e26a37f
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@d1c6f0b1487f90f107e9fd954f52daa15b6c373a
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@984de6fe7281eb715ad270245f7c6eb190f23def
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1576/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1576/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@58c377e943040d94f8b831c85ca1e91d4e26a37f
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@d1c6f0b1487f90f107e9fd954f52daa15b6c373a
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@984de6fe7281eb715ad270245f7c6eb190f23def
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

eatkins committed:
Simplify file management settings

I decided that there were too many settings related to the file
management that did similar things and had similar names but did
slightly different things. To improve this, I introduce the ChangedFiles
class to sbt.nio.file and switch to having just two task for file input
and output retrieval: all(Input|Output)Files and
changed(Input|Output)Files. If, for example, changedInputFiles returns
None that means that either the task has not yet been run or there were
no changes. If there have been any changes, then it will return
Some(changes) and the user can extract the relevant changes that they
are interested in.

The code may be slightly more verbose in a few places, but I think it's
worth it for the conceptual clarity.

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1582/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1582/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@58c377e943040d94f8b831c85ca1e91d4e26a37f
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@d1c6f0b1487f90f107e9fd954f52daa15b6c373a
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@984de6fe7281eb715ad270245f7c6eb190f23def
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1583/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1583/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@6d4a6457c35ff086df3eec97cfe30454cdb0384e
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@b630b5531d4554dcf6556970cea9cbeb3ff8a449
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@984de6fe7281eb715ad270245f7c6eb190f23def
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1586/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1586/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@2a4c7d1204c54c3d0050482196948136772e418e
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@b630b5531d4554dcf6556970cea9cbeb3ff8a449
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@984de6fe7281eb715ad270245f7c6eb190f23def
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1587/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1587/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@2a4c7d1204c54c3d0050482196948136772e418e
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@b630b5531d4554dcf6556970cea9cbeb3ff8a449
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@984de6fe7281eb715ad270245f7c6eb190f23def
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1588/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-pull/226/head-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1588/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@2a4c7d1204c54c3d0050482196948136772e418e
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/pull/226">pull/226/head       </A> | sbt/io@b630b5531d4554dcf6556970cea9cbeb3ff8a449
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@984de6fe7281eb715ad270245f7c6eb190f23def
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=pull/226/head&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

eed3si9n commented:
Rebuilding since IO 1.3.0-M8 is out.

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1589/console">is in progress...</a>

eed3si9n commented:
For the record, Dbuild failed with

```
[sbt212] [error] Error: Total 15, Failed 0, Errors 1, Passed 14
[sbt212] [error] Error during tests:
[sbt212] [error] 	sbt.internal.util.SettingsTest
```

```
[sbt212:error] java.lang.StackOverflowError
[sbt212:error] 	at sbt.internal.util.Init$Value.$anonfun$apply$3(Settings.scala:823)
[sbt212:error] 	at sbt.internal.util.Init$Value.$anonfun$apply$3(Settings.scala:823)
[sbt212:error] 	at sbt.internal.util.Init$Value.$anonfun$apply$3(Settings.scala:823)
[sbt212:error] 	at sbt.internal.util.Init$Value.$anonfun$apply$3(Settings.scala:823)
[sbt212:error] 	at sbt.internal.util.Init$Value.$anonfun$apply$3(Settings.scala:823)
[sbt212:error] 	at sbt.internal.util.Init$Value.$anonfun$apply$3(Settings.scala:823)
[sbt212:error] 	at sbt.internal.util.Init$Value.$anonfun$apply$3(Settings.scala:823)
```

eed3si9n commented:
AppVeyor failed on both Visual Studio 2015 and Visual Studio 2017


```
Running actions/clean-keep
[error] java.nio.file.FileSystemException: C:\Users\appveyor\AppData\Local\Temp\1\sbt_d62fda52\project\target\config-classes\$17a69f0703746b7bed64$.class: The process cannot access the file because it is being used by another process.
[error] 
[error] 	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:86)
[error] 	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97)
[error] 	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102)
[error] 	at sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:269)
[error] 	at sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:108)
[error] 	at java.nio.file.Files.deleteIfExists(Files.java:1165)
[error] 	at sbt.io.IO$.$anonfun$delete$3(IO.scala:559)
[error] 	at scala.collection.Iterator.foreach(Iterator.scala:941)
[error] 	at scala.collection.Iterator.foreach$(Iterator.scala:941)
[error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
[error] 	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
[error] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
[error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
[error] 	at sbt.io.IO$.$anonfun$delete$2(IO.scala:557)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] 	at sbt.internal.io.Retry$.liftedTree2$1(Retry.scala:38)
[error] 	at sbt.internal.io.Retry$.impl$1(Retry.scala:38)
[error] 	at sbt.internal.io.Retry$.apply(Retry.scala:50)
[error] 	at sbt.internal.io.Retry$.apply(Retry.scala:24)
[error] 	at sbt.io.IO$.delete(IO.scala:555)
[error] 	at sbt.io.IO$.$anonfun$delete$3(IO.scala:558)
[error] 	at scala.collection.Iterator.foreach(Iterator.scala:941)
[error] 	at scala.collection.Iterator.foreach$(Iterator.scala:941)
[error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
[error] 	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
[error] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
[error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
[error] 	at sbt.io.IO$.$anonfun$delete$2(IO.scala:557)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] 	at sbt.internal.io.Retry$.liftedTree2$1(Retry.scala:38)
[error] 	at sbt.internal.io.Retry$.impl$1(Retry.scala:38)
[error] 	at sbt.internal.io.Retry$.apply(Retry.scala:50)
[error] 	at sbt.internal.io.Retry$.apply(Retry.scala:24)
[error] 	at sbt.io.IO$.delete(IO.scala:555)
[error] 	at sbt.io.IO$.$anonfun$delete$3(IO.scala:558)
[error] 	at scala.collection.Iterator.foreach(Iterator.scala:941)
[error] 	at scala.collection.Iterator.foreach$(Iterator.scala:941)
[error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
[error] 	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
[error] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
[error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
[error] 	at sbt.io.IO$.$anonfun$delete$2(IO.scala:557)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] 	at sbt.internal.io.Retry$.liftedTree2$1(Retry.scala:38)
[error] 	at sbt.internal.io.Retry$.impl$1(Retry.scala:38)
[error] 	at sbt.internal.io.Retry$.apply(Retry.scala:50)
[error] 	at sbt.internal.io.Retry$.apply(Retry.scala:24)
[error] 	at sbt.io.IO$.delete(IO.scala:555)
[error] 	at sbt.io.IO$.$anonfun$delete$1(IO.scala:535)
[error] 	at sbt.io.IO$.$anonfun$delete$1$adapted(IO.scala:535)
[error] 	at scala.collection.Iterator.foreach(Iterator.scala:941)
[error] 	at scala.collection.Iterator.foreach$(Iterator.scala:941)
[error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
[error] 	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
[error] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
[error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
[error] 	at sbt.io.IO$.delete(IO.scala:535)
[error] 	at sbt.scriptedtest.ScriptedTests.$anonfun$runBatchedTests$1(ScriptedTests.scala:332)
...
[error] (scripted) java.nio.file.FileSystemException: C:\Users\appveyor\AppData\Local\Temp\1\sbt_d62fda52\project\target\config-classes\$17a69f0703746b7bed64$.class: The process cannot access the file because it is being used by another process.
```

eatkins commented:
Yeah. I don’t understand the dbuild failure yet because the test doesn’t fail on my computer. I’ll investigate the windows issue.

eed3si9n commented:
Travis failed only on header check:

```
[error] (mainProj / Test / headerCheck) There are files without headers!
[error]   /home/travis/build/sbt/sbt/main/src/test/scala/sbt/internal/FileStampJsonSpec.scala
```

eed3si9n commented:
@eatkins Maybe try https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Thread.html#yield() instead of `sleep(0)` in retry? On .NET for instance apparently they mean slightly different things.

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1589/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@2a4c7d1204c54c3d0050482196948136772e418e
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@984ec85b609664800366d375a49596487c7c3d9f
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@984de6fe7281eb715ad270245f7c6eb190f23def
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

eatkins commented:
I have a reproduction of the windows issue on my VM. I should likely be able to fix it by tomorrow.

eed3si9n commented:
For Dbuild stackoverflow, I am bumping up the stack to see if it helps - https://github.com/sbt/sbt-validator/pull/46

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1590/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/4627/head-develop-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1590/">has checked</A> the following projects, tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/4627">pull/4627/head      </A> | sbt/sbt@479c2c21a9c6222655fc11fbb0407b391ae2b6b7
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/develop">develop             </A> | sbt/zinc@b6fc35f27615f129f91de9ac14400101fb8b9b7e
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@984ec85b609664800366d375a49596487c7c3d9f
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@984de6fe7281eb715ad270245f7c6eb190f23def
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@3568b13a2e257f33957dd32c51c56ce9c378c9f9

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/4627/head&zinc=develop&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

eed3si9n commented on sbt/src/sbt-test/nio/make-clone/build.sbt:
```diff
@@ -0,0 +1,132 @@
+linkLib := {
+  val changedObjects = (compileLib / changedOutputFiles).value
+  val outPath = (linkLib / target).value.toPath
+  val allObjects = (compileLib / allOutputFiles).value.map(_.toString)
```
What's the difference between `compileLib` and `compileLib / allOutputFiles`?

dwijnand commented on main/src/main/scala/sbt/internal/SettingsGraph.scala:
```diff
@@ -14,38 +14,26 @@ import sbt._
-}
-private[sbt] object InputGraph {
-  @deprecated("Source is also deprecated.", "1.3.0")
+private[sbt] object SettingsGraph {
```
There already is a `SettingGraph`, so this is a little confusing...

eatkins replied:
I can’t argue with that. I am actually hoping/planning to get rid of this anyway. I understand the problem better than when I first introduced this.
